### PR TITLE
Fix #334

### LINKED ISSUE #334: Implement `java.lang.Object::clone`
As discovered by @sjrd we don't have the implementation for that yet.

### PR DESCRIPTION
**Pull request by LukasKellenberger**

### CODE REVIEW

**Comment by MasseGuillaume** [comment]
#334 Implement `java.lang.Object::clone` :-)

File: nativelib/src/main/scala/java/lang/Object.scala [hunk 1]
```diff
@@ -37,7 +39,17 @@ class _Object {
   @inline def __scala_## : scala.Int =
     __hashCode
 
-  protected def __clone(): _Object = ???
+  protected def __clone(): _Object = {
+    val typePtr: Ptr[Type] = getType(this)
+    val size               = (!typePtr).size
+    val clone              = GC.malloc_atomic(size)
```
Comment (densh) [inline nativelib/src/main/scala/java/lang/Object.scala]: Just `GC.malloc`, not `GC.malloc_atomic`. Atomic means that object stores no pointers to other objects.

File: nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
+package scala.scalanative.nir
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala]: It belongs to `scala.scalanative.optimizer.analysis`.

File: nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        Padding(finalPad, size) :: positionedTypes
+      } else {
+        positionedTypes
+      }
+
+    (potys.foldLeft(0L) { case (acc, poty) => acc + poty.size }, potys)
+  }
+}
+
+case class MemoryLayout(size: Long, tys: List[PositionedType])
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala]: `final case class`

File: nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala [hunk 4]
```diff
@@ -0,0 +1,70 @@
+package scala.scalanative.nir
+
+import MemoryLayout.PositionedType
+
+object MemoryLayout {
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala]: We put companion objects after the classes.

File: nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+  sealed abstract class PositionedType {
+    def size: Long
+    def offset: Long
+  }
+  case class Tpe(size: Long, offset: Long, ty: Type) extends PositionedType
+  case class Padding(size: Long, offset: Long)       extends PositionedType
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala]: `final case class`

File: nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+import MemoryLayout.PositionedType
+
+object MemoryLayout {
+
+  sealed abstract class PositionedType {
+    def size: Long
+    def offset: Long
+  }
+  case class Tpe(size: Long, offset: Long, ty: Type) extends PositionedType
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala]: `final case class`

File: nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  case class Padding(size: Long, offset: Long)       extends PositionedType
+
+  def apply(tys: Seq[Type]): MemoryLayout = {
+    val (size, potys) = helper(tys, 0)
+
+    MemoryLayout(size, potys.reverse)
+  }
+  private def helper(tys: Seq[Type],
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala]: `def impl`

File: nir/src/main/scala/scala/scalanative/nir/Types.scala [hunk 8]
```diff
@@ -17,6 +18,16 @@ sealed abstract class Type {
 
   final def show: String = nir.Show(this)
 
+  def size: Long = this match {
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Types.scala]: `final def` to make it clear that it's never overriden.

File: nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala (hunk 7 above)
Comment (LukasKellenberger) [inline nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala]: Should I remove `private` ?

Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala]: Sorry, it wasn't clear. `private` is fine, it's just that typically those kind of methods are called `impl`.

Comment (LukasKellenberger) [inline nir/src/main/scala/scala/scalanative/nir/MemoryLayout.scala]: Ok

**Comment by densh** [comment]
LGTM!

**Pull request merged**